source/connection_manager.c

```diff
@@ -0,0 +1,876 @@
+         * a callback.  So we need to re-lock and update the state ourselves.
+         */
+        aws_mutex_lock(&manager->lock);
+
+        AWS_FATAL_ASSERT(manager->pending_connects_count >= new_connection_failures);
+        manager->pending_connects_count -= new_connection_failures;
+
+        /*
+         * Rather than failing one acquisition for each connection failure, if there's at least one
+         * connection failure, we instead fail all excess acquisitions, since there's no pending
+         * connect that will necessarily resolve them.
+         *
+         * Try to correspond an error with the acquisition failure, but as a fallback just use the
+         * representative error.
+         */
+        size_t i = 0;
+        while (manager->pending_acquisition_count > manager->pending_connects_count) {
+            int error = representative_error;
+            if (i < aws_array_list_length(&errors)) {
+                aws_array_list_get_at(&errors, &error, i);
+            }
+
+            s_aws_http_connection_manager_move_front_acquisition(manager, NULL, error, completions);
+            ++i;
+        }
+
+        aws_mutex_unlock(&manager->lock);
+    }
+
+    s_aws_http_connection_manager_complete_acquisitions(completions, manager->allocator);
+
+    aws_array_list_clean_up(&errors);
+}
+
+int aws_http_connection_manager_acquire_connection(
+    struct aws_http_connection_manager *manager,
+    aws_http_on_client_connection_setup_fn *callback,
+    void *user_data) {
+
+    int result = AWS_OP_ERR;
+
+    struct aws_http_connection_acquisition *request =
+        aws_mem_acquire(manager->allocator, sizeof(struct aws_http_connection_acquisition));
+    if (request == NULL) {
+        callback(NULL, aws_last_error(), user_data);
+        return result;
+    }
+
+    AWS_LOGF_DEBUG(AWS_LS_HTTP_CONNECTION_MANAGER, "id=%p: Acquire connection", (void *)manager);
+
+    AWS_ZERO_STRUCT(*request);
+    request->callback = callback;
+    request->user_data = user_data;
+
+    struct aws_linked_list completions;
+    aws_linked_list_init(&completions);
+
+    size_t new_connections = 0;
+
+    aws_mutex_lock(&manager->lock);
+
+    AWS_ASSERT(manager->state == AWS_HCMST_READY);
+    if (manager->state == AWS_HCMST_READY) {
+        aws_linked_list_push_back(&manager->pending_acquisitions, &request->node);
+        ++manager->pending_acquisition_count;
+
+        s_aws_http_connection_manager_build_task_set(manager, &completions, &new_connections);
+        result = AWS_OP_SUCCESS;
+    } else {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_CONNECTION_MANAGER,
+            "id=%p: Acquire connection called when manager in shut down state",
+            (void *)manager);
+
+        request->error_code = AWS_ERROR_HTTP_CONNECTION_MANAGER_INVALID_STATE_FOR_ACQUIRE;
+        aws_linked_list_push_back(&completions, &request->node);
+
+        aws_raise_error(AWS_ERROR_HTTP_CONNECTION_MANAGER_INVALID_STATE_FOR_ACQUIRE);
```

Comment:
The completion callback shouldn't fire if aws_http_connection_manager_acquire_connection() returns an error code. I can't tell whether or not this is the case...

That's just the pattern we've been using. If error returned, then no completion callback.

Reply:
Ok, yeah, it's definitely calling the callback on several different synchronous errors.  I'll remove.

Reply:
Actually scratch that.  Per discussion, we'll remove the return value on this function and instead always notify on the callback.